ospd_openvas/messaging/mqtt.py

```diff
@@ -0,0 +1,159 @@
+        bind_address="",
+        bind_port=0,
+        clean_start=mqtt.MQTT_CLEAN_START_FIRST_ONLY,
+        properties=None,
+    ):
+        if not host:
+            host = self._mqtt_broker_address
+        if not port:
+            port = self._mqtt_broker_port
+
+        return super().connect(
+            host,
+            port=port,
+            keepalive=keepalive,
+            bind_address=bind_address,
+            bind_port=bind_port,
+            clean_start=clean_start,
+            properties=properties,
+        )
+
+
+class MQTTPublisher(Publisher):
+    def __init__(self, client: MQTTClient):
+        self._client = client
+
+    def publish(self, message: Message) -> None:
+        logger.debug('Publish message %s', message)
+        self._client.publish(message.topic, str(message), qos=QOS_AT_LEAST_ONCE)
+
+
+class MQTTSubscriber(Subscriber):
+    def __init__(self, client: MQTTClient):
+        self._client = client
+
+    def subscribe(
+        self, message_class: Type[Message], callback: Callable[[Message], None]
+    ) -> None:
+        func = partial(self._handle_message, message_class, callback)
+        func.__name__ = callback.__name__
+
+        logger.debug("Subscribing to topic %s", message_class.topic)
+
+        self._client.subscribe(message_class.topic, qos=QOS_AT_LEAST_ONCE)
+        self._client.message_callback_add(message_class.topic, func)
+
+    @staticmethod
+    def _handle_message(
+        message_class: Type[Message],
+        callback: Callable[[Message], None],
+        _client,
+        _userdata,
+        msg: mqtt.MQTTMessage,
+    ) -> None:
+        logger.debug("Incoming message for topic %s", msg.topic)
+
+        try:
+            # Load message from payload
+            message = message_class.load(msg.payload)
+        except json.JSONDecodeError:
+            logger.error(
+                "Got MQTT message in non-json format for topic %s.", msg.topic
+            )
+            logger.debug("Got: %s", msg.payload)
+            return
+        except ValueError as e:
+            logger.error(
+                "Could not parse message for topic %s. Error was %s",
+                msg.topic,
+                e,
+            )
+            logger.debug("Got: %s", msg.payload)
+            return
+
+        callback(message)
+
+
+class MQTTDaemon:
+    """A class to start and stop the MQTT client"""
+
+    def __init__(
+        self,
+        client: MQTTClient,
+    ):
+        self._client = client
+        self._client.on_disconnect = self.on_disconnect
+        self._client.on_connect = self.on_connect
+
+        self._client.connect()
+
+    @staticmethod
+    def on_connect(_client, _userdata, _flags, rc, _properties):
+        if rc == 0:
+            logger.info("Connected to broker successfully")
+        else:
+            logger.error('Failed to connect to broker. Reason code %s', rc)
+
+    @staticmethod
+    def on_disconnect(client, _userdata, rc=0):
```

Comment:
```suggestion
    def on_disconnect(client: MQTTClient, _userdata, rc: int = 0) -> None:
```
I am not sure what userdata is for, so you may fill it on your own ...?

Reply:
userdata can be anything or nothing, you can give the function any additional information you like within the client itself. You can set it with something like
```py
self._client.user_data_set()
```
or
```py
self._client._userdata = "foo"
```
therefore I don't think it makes any sense to set any type for it. Also I thing it is not necessary to specify any type to the `on_`-functions as they are not intended to be called by a user.

Reply:
yeah. See my edit ... I looked into paho ...